neuromation/cli/ael.py

```diff
@@ -0,0 +1,460 @@
+    "==================== job's logs ====================\n", dim=True
+)
+ATTACH_STARTED = click.style(
+    "=================== job's output ===================\n", dim=True
+)
+ATTACH_STARTED_AFTER_LOGS = click.style(
+    "======== job's output, may overlap with logs =======\n", dim=True
+)
+
+
+@dataclasses.dataclass(frozen=True)
+class AttachHelper:
+    attach_ready: asyncio.Event
+    log_printed: asyncio.Event
+    write_sem: asyncio.Semaphore
+
+
+async def process_logs(root: Root, job: str, helper: Optional[AttachHelper]) -> None:
+    codec_info = codecs.lookup("utf8")
+    decoder = codec_info.incrementaldecoder("replace")
+    async for chunk in root.client.jobs.monitor(job):
+        if not chunk:
+            break
+        txt = decoder.decode(chunk)
+        if helper is not None:
+            if helper.attach_ready.is_set():
+                return
+            async with helper.write_sem:
+                if not helper.log_printed.is_set() and not root.quiet:
+                    click.echo(LOGS_STARTED, nl=False)
+                helper.log_printed.set()
+                click.echo(txt, nl=False)
+        else:
+            click.echo(txt, nl=False)
+
+
+async def process_exec(root: Root, job: str, cmd: str, tty: bool) -> None:
+    exec_id = await root.client.jobs.exec_create(job, cmd, tty=tty)
+    if tty:
+        await _exec_tty(root, job, exec_id)
+    else:
+        await _exec_non_tty(root, job, exec_id)
+
+    info = await root.client.jobs.exec_inspect(job, exec_id)
+    progress = ExecStopProgress.create(tty=root.tty, color=root.color, quiet=root.quiet)
+    while info.running:
+        await asyncio.sleep(0.1)
+        info = await root.client.jobs.exec_inspect(job, exec_id)
+        if not progress():
+            sys.exit(EX_IOERR)
+    sys.exit(info.exit_code)
+
+
+async def _exec_tty(root: Root, job: str, exec_id: str) -> None:
+    loop = asyncio.get_event_loop()
+    stdout = create_output()
+    h, w = stdout.get_size()
+    async with root.client.jobs.exec_start(job, exec_id) as stream:
+        try:
+            await root.client.jobs.exec_resize(job, exec_id, w=w, h=h)
+        except IllegalArgumentError:
+            info = await root.client.jobs.exec_inspect(job, exec_id)
+            if not info.running:
+                # Exec session is finished
+                return
+
+        tasks = []
+        tasks.append(loop.create_task(_process_stdin_tty(stream)))
+        tasks.append(loop.create_task(_process_stdout_tty(stream, stdout)))
+        tasks.append(
+            loop.create_task(
+                _process_resizing(
+                    functools.partial(root.client.jobs.exec_resize, job, exec_id),
+                    stdout,
+                )
+            )
+        )
+        await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+        for task in tasks:
+            await root.cancel_with_logging(task)
+
+
+async def _exec_non_tty(root: Root, job: str, exec_id: str) -> None:
```

Comment:
Should not we read stdin even in non-tty mode?

Reply:
Excellent question!
Technically it is the possible scenario (while quite rare).
Let me briefly check how to do it in OS-agnostic manner.

Reply:
done